### PR TITLE
Add state checkpointing and uploading in coordinator

### LINKED ISSUE #237: Keep the latest checkpoint on non-gpu peer(s)
Currently, GPU peers save the latest model checkpoint locally, preventing which is problematic if they are running on preemptible vm/colab instances. It would be great to make `run_first_peer.py` into a checkpoint holder

- [ ] Please add an option for `run_first_peer.py` to periodically load checkpoints
   - [ ] make sure that if all peers except for the first one will leave AND then then new peers will join, they will download the checkpoint from the first peer
- [ ] Add an optional flag to upload the latest model checkpoint to HF model hub
- [ ] as a side-quest, please also add `--initial_peers` parameter to `run_first_peer.py`

### PR DESCRIPTION
**Pull request by leshanbog**
Resolves #237

### CODE REVIEW

**Review by yhn112 — APPROVED** [review]
LGTM

**Review by justheuristic — APPROVED** [review]

File: examples/albert/run_first_peer.py [hunk 1]
```diff
@@ -1,51 +1,158 @@
...
 from whatsmyip.ip import get_ip
 
 import hivemind
+import torch
+from torch_optimizer import Lamb
+from transformers import AlbertForPreTraining, AlbertConfig, HfArgumentParser
 from hivemind.utils.logging import get_logger
 
+from run_trainer import DatasetArguments, CollaborationArguments, \
+    get_model, get_optimizer_and_scheduler
```
Comment (mryab) [inline examples/albert/run_first_peer.py:18]: Unused imports

File: examples/albert/run_first_peer.py [hunk 2]
```diff
@@ -1,51 +1,158 @@
...
 from whatsmyip.providers import GoogleDnsProvider
 from whatsmyip.ip import get_ip
 
 import hivemind
+import torch
+from torch_optimizer import Lamb
+from transformers import AlbertForPreTraining, AlbertConfig, HfArgumentParser
 from hivemind.utils.logging import get_logger
 
+from run_trainer import DatasetArguments, CollaborationArguments, \
```
Comment (mryab) [inline examples/albert/run_first_peer.py:17]: If these functions are used more than in one script, it's best to extract them to a separate module

File: examples/albert/run_first_peer.py [hunk 3]
```diff
@@ -1,51 +1,158 @@
...
+        config = AlbertConfig.from_pretrained(dataset_args.config_path)
+        self.model = AlbertForPreTraining(config)
+
+        collaboration_args_dict = asdict(coordinator_args)
+        collaboration_args_dict.pop('address')
+        collaboration_args_dict.pop('refresh_period')
+        collaboration_args_dict.pop('wandb_project')
+        collaboration_args_dict.pop('save_checkpoint_step_interval')
+        collaboration_args_dict.pop('upload_model_as')
+        collaboration_args_dict.pop('upload_interval')
```
Comment (mryab) [inline examples/albert/run_first_peer.py:54]: You inherit from CollaborationArguments only to pop all the arguments afterwards :) Why not create a separate group of arguments, so that you won't have to write these lines?

File: examples/albert/run_first_peer.py [hunk 4]
```diff
@@ -1,51 +1,158 @@
...
+        )
+
+        adjusted_target_batch_size = collaboration_args_dict.pop('target_batch_size') - \
+            collaboration_args_dict.pop('batch_size_lead')
+
+        collaboration_args_dict.pop('initial_peers')
+        collaboration_args_dict.pop('trainer_uuid')
+        collaboration_args_dict.pop('dht_listen_on')
+        collaboration_args_dict.pop('statistics_expiration')
+        collaboration_args_dict.pop('endpoint')
```
Comment (mryab) [inline examples/albert/run_first_peer.py:80]: Dropping all these parsed arguments leads me to the thought that their hierarchy can be improved. Is it possible to extract all common arguments for peers and collaborators to one argument group and then inherit from it in two separate classes?

File: examples/albert/run_first_peer.py [hunk 5]
```diff
@@ -1,51 +1,158 @@
...
+        else:
+            return False
+
+    def upload_checkpoint(self):
+        self.model.save_pretrained(self.upload_model_as)
+        torch.save(self.collaborative_optimizer.opt.state_dict(), f"{self.upload_model_as}/optimizer_state.pt")
+        self.previous_timestamp = time.time()
+        try:
+            subprocess.run("git add --all".split(), check=True, cwd=self.upload_model_as)
+            subprocess.run("git commit -m Updating".split(), check=True, cwd=self.upload_model_as)
```
Comment (mryab) [inline examples/albert/run_first_peer.py:117]: I'd like to have a more informative commit message :) For instance, report the step number and the training loss

File: examples/albert/run_first_peer.py [hunk 6]
```diff
@@ -1,51 +1,158 @@
...
+    address: Optional[str] = None  # this machine's network address. Use public IP for global experiments,
+    # local address for private runs
+    refresh_period: float = 30  # coordinator will fetch keys from DHT once in this many seconds
+    wandb_project: Optional[str] = None  # learning curves will be published there
+    save_checkpoint_step_interval: int = 5  # coordinator will load and save state from peers once every that many steps
+    upload_model_as: Optional[str] = None  # coordinator will upload the checkpoint to that HuggingFace repo
+    upload_interval: Optional[float] = None  # coordinator will upload model once in this many seconds
+    # Note: You might want to have several initial peers so that if one dies,
+    # new workers still can join the collaboration via alive initial peers' addresses.
+    # Specify initial_peers argument for that purpose
```
Comment (mryab) [inline examples/albert/run_first_peer.py:35]: By the way, can the description of these arguments be written in the same way as transformers TrainingArguments? I mean mainly the metadata "help" field: we shouldn't force the user to read the code in order to understand the meaning of each argument

File: examples/albert/run_first_peer.py [hunk 7]
```diff
@@ -1,51 +1,158 @@
...
+            return True
+        else:
+            return False
+
+    def upload_checkpoint(self):
+        self.model.save_pretrained(self.upload_model_as)
+        torch.save(self.collaborative_optimizer.opt.state_dict(), f"{self.upload_model_as}/optimizer_state.pt")
+        self.previous_timestamp = time.time()
+        try:
+            subprocess.run("git add --all".split(), check=True, cwd=self.upload_model_as)
```
Comment (mryab) [inline examples/albert/run_first_peer.py:116]: 1) .split() is unnecessary if you pass shell=True
2) Is the current working directory of a process the same thing as the Model Hub repository name?

File: examples/albert/run_first_peer.py [hunk 8]
```diff
@@ -1,51 +1,158 @@
...
+            subprocess.run("git commit -m Updating".split(), check=True, cwd=self.upload_model_as)
+            subprocess.run("git push".split(), check=True, cwd=self.upload_model_as)
+        except subprocess.CalledProcessError:
+            logger.warning("Error while uploading model")
```
Comment (mryab) [inline examples/albert/run_first_peer.py:120]: You can [catch and report](https://stackoverflow.com/a/24850026) the exact error

File: examples/albert/run_first_peer.py [hunk 9]
```diff
@@ -1,51 +1,158 @@
...
+    save_checkpoint_step_interval: int = 5  # coordinator will load and save state from peers once every that many steps
+    upload_model_as: Optional[str] = None  # coordinator will upload the checkpoint to that HuggingFace repo
+    upload_interval: Optional[float] = None  # coordinator will upload model once in this many seconds
+    # Note: You might want to have several initial peers so that if one dies,
+    # new workers still can join the collaboration via alive initial peers' addresses.
+    # Specify initial_peers argument for that purpose
+
+
+class CheckpointHandler:
+    def __init__(self, coordinator_args: CoordinatorArguments, dataset_args: DatasetArguments, dht: hivemind.DHT):
```
Comment (mryab) [inline examples/albert/run_first_peer.py:39]: Receiving the entire DatasetArguments object only to get a config path is probably overkill

File: examples/albert/run_trainer.py [hunk 10]
```diff
@@ -29,8 +29,9 @@ class CollaborationArguments:
     """ define how peers interact with each other while training"""
 
     # primary parameters
-    initial_peers: List[str]  # one or more peers (comma-separated) that will welcome you into the collaboration
     experiment_prefix: str  # a unique "name" of this experiment, used to store metadata on the DHT
+    initial_peers: List[str] = field(default_factory=list)  # one or more peers (comma-separated)
+    # that will welcome you into the collaboration
```
Comment (mryab) [inline examples/albert/run_trainer.py:34]: Here, it'd also be great if we provided help for all arguments in the CLI

**Review by mryab — CHANGES_REQUESTED** [review]

File: examples/albert/run_first_peer.py [hunk 11]
```diff
@@ -1,45 +1,159 @@
...
+        default=5,
+        metadata={"help": "Coordinator will load and save state from peers once every that many steps"}
+    )
+    model_config_path: str = field(
+        default='https://s3.amazonaws.com/models.huggingface.co/bert/albert-large-v2-config.json',
+        metadata={"help": "Path to the model config"}
+    )
+    repo_path: Optional[str] = field(
+        default=None,
+        metadata={"help":  "Path to HuggingFace repo in which coordinator will upload the model and optimizer states"}
```
Comment (mryab) [inline examples/albert/run_first_peer.py:47]: Two spaces after the colon; make sure to check the formatting

File: examples/albert/run_first_peer.py [hunk 12]
```diff
@@ -1,45 +1,159 @@
...
-    parser.add_argument('--experiment_prefix', type=str, required=True,
-                        help="a prefix where peers store their metrics for aggregation")
-    parser.add_argument('--wandb_project', type=str, required=True,
-                        help="Weights & Biases project name to publish learning curves")
-
-    args = parser.parse_args()
-    if args.address is None:
+    parser = HfArgumentParser(CoordinatorArguments)
+    coordinator_args = parser.parse_args_into_dataclasses()[0]
+    coordinator_args_dict = asdict(coordinator_args)
```
Comment (mryab) [inline examples/albert/run_first_peer.py:134]: When I was suggesting that we separate the arguments into groups, I also meant that we can parse them separately into several groups of arguments (see examples in the Transformers library or even in the run_trainer script)

This way, you 
1) won't have to repeatedly call pop() for the arguments, because they won't be passed into wrong classes
2) ideally, won't have to use asdict() unless the very moment when you're initializing internal Hivemind classes

File: hivemind/optim/collaborative.py [hunk 13]
```diff
@@ -208,7 +210,8 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
             self.collaboration_state_updated.set()
             self.update_scheduler()
 
-            logger.log(self.status_loglevel, f"Optimizer step: done!")
+            logger.log(self.status_loglevel, f"Optimizer step: done! "
+                                             f"Your current contribution: {self.samples_processed} samples")
```
Comment (mryab) [inline hivemind/optim/collaborative.py:214]: I'd prefer to report the current contribution on the INFO level repeatedly regardless of the verbosity level, but for now, it's probably OK as is

**Review by mryab — CHANGES_REQUESTED** [review]

File: examples/albert/run_first_peer.py [hunk 14]
```diff
@@ -1,45 +1,158 @@
...
+            return False
+
+    def upload_checkpoint(self, current_loss):
+        self.model.save_pretrained(self.repo_path)
+        torch.save(self.collaborative_optimizer.opt.state_dict(), f"{self.repo_path}/optimizer_state.pt")
+        self.previous_timestamp = time.time()
+        try:
+            subprocess.run("git add --all", shell=True, check=True, cwd=self.repo_path)
+            current_step = self.collaborative_optimizer.collaboration_state.optimizer_step
+            subprocess.run(f"git commit -m 'Step {current_step}, loss {round(current_loss, 3)}'",
```
Comment (mryab) [inline examples/albert/run_first_peer.py:127]:
```suggestion
            subprocess.run(f"git commit -m 'Step {current_step}, loss {current_loss:.3f}'",
```

File: examples/albert/run_first_peer.py [hunk 15]
```diff
@@ -58,12 +171,17 @@
...
-                })
-                logger.info(f"Step #{current_step}\tloss = {sum_loss / alive_peers:.5f}")
+                if coordinator_args.wandb_project is not None:
+                    wandb.log({
+                        "loss": sum_loss / sum_mini_steps,
+                        "alive peers": alive_peers,
+                        "samples": num_samples,
+                        "performance": sum_perf
+                    })
+                logger.info(f"Step #{current_step}\tloss = {sum_loss / sum_mini_steps:.5f}")
```
Comment (mryab) [inline examples/albert/run_first_peer.py:181]:
```suggestion
                logger.info(f"Step #{current_step}\tloss = {sum_loss / sum_mini_steps:.3f}")
```

**Review by mryab — APPROVED** [review]

**Comment by borzunov** [comment]
@leshanbog Please remove `trainer_uuid` from `CollaborationArguments`, as the merged commit from the `master` branch did :)
